sdk/storage/Azure.Storage.Blobs/src/BlobDirectoryClient.cs

```diff
@@ -0,0 +1,726 @@
+        {
+            _uri = blobUri;
+            if (!string.IsNullOrEmpty(blobUri.Query))
+            {
+                UriQueryParamsCollection queryParamsCollection = new UriQueryParamsCollection(blobUri.Query);
+            }
+
+            _clientConfiguration = clientConfiguration;
+            _clientSideEncryption = clientSideEncryption?.Clone();
+
+            BlobErrors.VerifyHttpsCustomerProvidedKey(_uri, _clientConfiguration.CustomerProvidedKey);
+            BlobErrors.VerifyCpkAndEncryptionScopeNotBothSet(_clientConfiguration.CustomerProvidedKey, _clientConfiguration.EncryptionScope);
+        }
+        #endregion
+
+        /// <summary>
+        /// Intended to create a backing blob client for data movement.
+        /// Looking to remove this if it's unnecessary to do this.
+        ///
+        /// Initializes a new instance of the <see cref="BlobServiceClient"/>
+        /// class.
+        /// </summary>
+        /// <returns>
+        /// New instanc of the <see cref="BlobServiceClient"/> class.
+        /// </returns>
+        public BlobDirectoryClient CloneClient()
+        {
+            return new BlobDirectoryClient(Uri, ClientConfiguration, ClientSideEncryption);
+        }
+
+        #region Upload
+        /// <summary>
+        /// The <see cref="Upload(string, BlobUploadDirectoryOptions, CancellationToken)"/>
+        /// operation overwrites the contents of the blob directory, creating a new blob
+        /// if none exists.  Overwriting an existing block blob replaces
+        /// any existing metadata on the blob.
+        ///
+        /// Set <see href="https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-conditional-headers-for-blob-service-operations">
+        /// access conditions</see> through <see cref="BlobUploadOptions.Conditions"/>
+        /// to avoid overwriting existing data.
+        ///
+        /// For now this will only do block blobs. For the future, we can have in BlobUploadDirectoryOptions
+        ///
+        /// For more information, see
+        /// <see href="https://docs.microsoft.com/rest/api/storageservices/put-blob">
+        /// Put Blob</see>.
+        /// </summary>
+        /// <param name="directory">
+        /// A <see cref="Directory"/> containing the content to upload.
+        /// </param>
+        /// <param name="options">
+        /// Optional parameters.
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response{BlobContentInfo}"/> describing the
+        /// state of the updated block blob.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual Response<BlobContentInfo> Upload(
```

Comment:
@rnpmsft  is working on changing this so that the return type will contain the results of all the blobs in the directory.

For the callback, I was thinking that could be something in the progressHandler in the BlobUploadDirectoryOptions they could specify.

Reply:
Will be done in a separate PR by Rushi.